libbeat/processors/convert/convert.go

```diff
@@ -0,0 +1,299 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package convert
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+const logName = "processor.convert"
+
+func init() {
+	processors.RegisterPlugin("convert", New)
+}
+
+type processor struct {
+	config
+	log *logp.Logger
+}
+
+// New constructs a new convert processor.
+func New(cfg *common.Config) (processors.Processor, error) {
+	c := defaultConfig()
+	if err := cfg.Unpack(&c); err != nil {
+		return nil, errors.Wrap(err, "fail to unpack the convert processor configuration")
+	}
+
+	return newConvert(c)
+}
+
+func newConvert(c config) (*processor, error) {
+	log := logp.NewLogger(logName)
+	if c.Tag != "" {
+		log = log.With("instance_id", c.Tag)
+	}
+
+	return &processor{config: c, log: log}, nil
+}
+
+func (p *processor) String() string {
+	return fmt.Sprintf("convert=[fields=%v, ignore_failure=%v, ignore_missing=%v, instance_id=%v, mode=%v]",
+		p.Fields, p.IgnoreFailure, p.IgnoreMissing, p.Tag, p.Mode)
+}
+
+func (p *processor) Run(event *beat.Event) (*beat.Event, error) {
+	for _, conv := range p.Fields {
+		v, _ := event.GetValue(conv.From)
+		if v == nil {
+			if !p.IgnoreMissing {
+				return event, annotateError(p.Tag, errors.Errorf("field [%v] is missing, cannot be converted to type [%v]", conv.From, conv.Type))
+			}
+			continue
+		}
+
+		if conv.Type > unset {
+			t, err := p.transformType(conv.Type, v)
+			if err != nil {
+				if !p.IgnoreFailure {
+					return event, annotateError(p.Tag, errors.Wrapf(err, "unable to convert field [%v] value [%v] to [%v]", conv.From, v, conv.Type))
```

Comment:
Right now on failure the event might be already modified by previous fields. Is this expected or should it reset to the original event? I think we do that in other processors. Same above with ignore missing.

Reply:
@ruflin I have finally addressed this and added a few tests to check that the event is not modified at all if there is a failure.